File: leetcode/Tree/非递归遍历二叉树/530. 二叉搜索树的最小绝对差(中序).py
"""
给你一棵所有节点为非负值的二叉搜索树，请你计算树中任意两节点的差的绝对值的最小值。

 

示例：

输入：

   1
    \
     3
    /
   2

输出：
1

解释：
最小绝对差为 1，其中 2 和 1 的差的绝对值为 1（或者 2 和 3）。
"""

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def getMinimumDifference(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        # 找到两个相邻节点的差的最小值
        res = float("inf")
        pre = float("-inf")
        stack = []
        cur = root

        while cur or stack:
            while cur:
                stack.append(cur)
                cur = cur.left

            cur = stack.pop(-1)
            if cur.val - pre < res:
                res = cur.val - pre

            pre = cur.val

            cur = cur.right
        return res

File: lc/Tree/古城-树的遍历/非递归遍历二叉树/530. 二叉搜索树的最小绝对差(中序).go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func getMinimumDifference(root *TreeNode) int {

    var stack []*TreeNode
    pre := math.MinInt32
    res := math.MaxInt32

    for root != nil || len(stack) > 0 {
        for root != nil {
            stack = append(stack, root)
            root = root.Left
        }

        // get last element
        cur := stack[len(stack) - 1]
        // remove last element
        stack = stack[:len(stack)-1]
        res = min(res, cur.Val - pre)
        pre = cur.Val

        root = cur.Right
    }

    return res
}